src/TzBot/TimeReference.hs

```
@@ -107,40 +92,33 @@ timeReferenceToUTC
   -> TimeReference -- ^ A time reference to translate to UTC.
   -> TimeReferenceToUTCResult
 timeReferenceToUTC sendersTZLabel eventTimestamp TimeReference {..} =
-  case mbEitherTzOrOffset of
+  case eithTzInfo of
     Left abbrev -> TRTUInvalidTimeZoneAbbrev abbrev
-    Right (Right offset) -> do
-      -- In the case of rigid offset we don't need the `modifyLocal` from
-      -- from the `tztime` package because there are no timeshifts that
-      -- we should take into account. So we just use plain LocalTime.
-      let refTime = eventTimestamp & (
-            convertUtcToOffsetTime offset
-            >>> dayTransition
-            >>> TZT.atTimeOfDay trTimeOfDay
-            )
-      TRTUSuccess $ TimeRefSuccess
-        (convertOffsetTimeToUtc offset refTime)
-        (Right offset)
-        (localDay refTime)
-    Right (Left (tzLabel, implicitSenderTimezone)) -> do
-      let eventLocalTime = TZT.fromUTC (TZI.fromLabel tzLabel) eventTimestamp
+    Right tzInfo -> do
+      let eventLocalTime = TZT.fromUTC tzInfo eventTimestamp
       let eithRefTime = eventLocalTime & TZT.modifyLocalStrict (
             dayTransition >>> TZT.atTimeOfDay trTimeOfDay
             )
       case eithRefTime of
-        Left err -> tzErrorToResult implicitSenderTimezone tzLabel err
+        Left err -> do
+          let mbRefTzLabel = case trLocationRef of
+                Just (TimeZoneRef refTzLabel) -> Just refTzLabel
+                Nothing -> Just sendersTZLabel
+                _ -> Nothing
+          case mbRefTzLabel of
+            Nothing -> error "impossible happened: got TZError for a static offset"
```

We don't really need this `error` branch if we change `TimeShiftErrorInfo.tseiRefTimeZone` from a `TZLabel` to a `TZInfo`.

Maybe I'm not explaining myself correctly; I pushed a demo of what I mean to `diogo/#67-refactor-timeReferenceToUTC`, see commit `991a597c1d907e3ebbcd76147eeb1e33abd1fb5a`

---

Well, I understood you correctly, but preferred to emit `error` explicitly rather than silently pass artificial `TZInfo` to the rendering stage, because in fact two behaviors are equivalent, but explicit `error` makes the code just more readable I believe.